### PR TITLE
Fixes #2441 - Make TaskReplaceActor wait after killing

### LINKED ISSUE #2441: AppRestart deployments don't wait for old tasks to be killed
AppRestart (TaskReplaceActor) deployments issue kills and then finish as soon as all new task are running, leaving outstanding kills.

As soon as the deployment is finished, the lock on the app is released and Marathon's normal scale code picks up. It usually notices that there are too many tasks running and, because it is not aware of the outstanding kills, it issues even more kills, making the running task count drop unnecessarily.

### PR DESCRIPTION
**Pull request by gkleiman**

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala [hunk 1]
```diff
@@ -113,36 +120,59 @@ class TaskReplaceActor(
...
+    periodicalRetryKills = context.system.scheduler.scheduleOnce(15.seconds, self, RetryKills)
 }
 
-private object ReplaceErrorState {
-  def unapply(state: String): Option[String] = state match {
-    case "TASK_ERROR" | "TASK_FAILED" | "TASK_KILLED" | "TASK_LOST" => Some(state)
-    case _ => None
-  }
+object TaskReplaceActor {
+  val TaskFinished = "^TASK_(ERROR|FAILED|FINISHED|LOST|KILLED)$".r
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala]: Rename this property. Irritating that TASK_XXX is TaskFinished where XXX can be FINISHED or something else.

File: src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala [hunk 2]
```diff
@@ -113,36 +120,59 @@ class TaskReplaceActor(
...
+  val TaskFailed = "^TASK_(ERROR|FAILED|LOST|KILLED)$".r
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala]: Rename this property. Irritating that TASK_XXX is TaskFailed where XXX can be FAILED or something else.

**Comment by gkleiman** [comment]
Comments addresed

File: src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala [hunk 3]
```diff
@@ -113,36 +122,59 @@ class TaskReplaceActor(
...
+    }
   }
 
   def buildTaskId(id: String): TaskID =
     TaskID.newBuilder()
       .setValue(id)
       .build()
+
+  def scheduleRetryKills(): Unit =
+    periodicalRetryKills = context.system.scheduler.scheduleOnce(15.seconds, self, RetryKills)
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala]: can you add a cancel of that timer in postStop - just for safety?

File: src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala [hunk 4]
```diff
@@ -87,25 +89,32 @@ class TaskReplaceActor(
   }
 
   def commonBehavior: PartialFunction[Any, Unit] = {
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala]: Make the ReturnType `Receive`

File: src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala [hunk 5]
```diff
@@ -113,36 +122,59 @@ class TaskReplaceActor(
 
   def handleNewHealthyTask(taskId: String): Unit = {
     healthy += taskId
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala]: rename healthy to started? (can be either healthy or started)

**Comment by gkleiman** [comment]
Addressed the rest of the commits.

File: src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala [hunk 6]
```diff
@@ -29,6 +32,8 @@ class TaskReplaceActor(
   var oldTaskIds = tasksToKill.map(_.getId)
   val toKill = oldTaskIds.to[mutable.Queue]
   var maxCapacity = (app.instances * (1 + app.upgradeStrategy.maximumOverCapacity)).toInt
+  var outstandingKills = Set.empty[String]
+  var periodicalRetryKills: Cancellable = _
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala]: Can you make that an option - it really is one.

**Comment by aquamatthias** [comment]
I like it! Thanks!

**Pull request merged**